ddrop/layers.py

```diff
@@ -0,0 +1,37 @@
+from keras.layers import Dense
+from keras.wrappers import Wrapper
+import keras.backend as K
+
+
+class DropConnectDense(Dense):
+    def __init__(self, *args, **kwargs):
+        self.prob = kwargs.pop('prob', 0.5)
+        if 0. < self.prob < 1.:
+            self.uses_learning_phase = True
+        super(DropConnectDense, self).__init__(*args, **kwargs)
+
+    def call(self, x, mask=None):
+        if 0. < self.prob < 1.:
+            self.W = K.in_train_phase(K.dropout(self.W, self.prob), self.W)
+            self.b = K.in_train_phase(K.dropout(self.b, self.prob), self.b)
+
+        # Same as original
+        output = K.dot(x, self.W)
+        if self.bias:
+            output += self.b
+        return self.activation(output)
+
+
+class DropConnect(Wrapper):
+    def __init__(self, layer, prob=1., **kwargs):
+        self.prob = prob
+        self.layer = layer
+        if 0. < self.prob < 1.:
+            self.uses_learning_phase = True
+        super(DropConnect, self).__init__(**kwargs)
+
+    def call(self, x, mask=None):
+        if 0. < self.prob < 1.:
+            self.W = K.in_train_phase(K.dropout(self.W, self.prob), self.W)
```

Comment:
This is an approximation, because it scales it by (1-p) - see [keras/backend/theano_backend.py#L1303](https://github.com/fchollet/keras/blob/master/keras/backend/theano_backend.py#L1303).

To properly implement DropConnect, I was thinking that one cheap hack would be to multiply it by (1-p).

Also, you probably want to make use of `K.in_test_phase` to do the averaging thing during prediction... the list of operators available can be found at https://keras.io/backend/. You can also use TensorFlow operators directly if you don't mind breaking compatibility with Theano.